proposals/0370-modifiers.rst

```diff
@@ -122,39 +126,136 @@ Proposed Change Specification
     modifier may have a top-level type signature.
 
-12. A modifier of type ``Multiplicity`` changes the multiplicity of the following arrow,
-    or following pattern-bound variable of a lambda,
-    or preceding record field.
-    Multiple modifiers of type ``Multiplicity`` on the same arrow are not allowed.
-    Any other use of a modifier is an error.
-
-13. ``-XLinearTypes`` implies ``-XModifiers``.
-
-14. Future modifiers will be put *before* the element they modify. Alternatively,
+12. Future modifiers will be put *before* the element they modify. Alternatively,
     a modifier can be put directly before a syntactic closer or separator, such
     as ``;`` or ``where`` or ``)``.
 
-15. Modifiers with an unknown meaning produce a warning, controlled by
-    ``-Wunknown-modifiers``. They are otherwise ignored. (However, in order to
-    know that a modifier is unknown, it still must be parsed, renamed, and type-checked.)
+13. Modifiers of unknown kind produce an error.
+
+14. Modifiers of known kind but with an unknown meaning produce a warning,
+    controlled by ``-Wunknown-modifiers``. They are otherwise ignored. (However,
+    in order to know that a modifier is unknown, it still must be parsed,
+    renamed, and type-checked.)
+
+LinearTypes
+-----------
+With ``-XLinearTypes``:
+
+* A modifier of type ``Multiplicity`` changes the multiplicity of the following
+  arrow, or following pattern-bound variable of a lambda, or following let or
+  where binding, or preceding record field. Multiple modifiers of type
+  ``Multiplicity`` on the same arrow are not allowed. Any other use of a
+  modifier still has no meaning.
+
+* The ``%1`` modifier is handled as a special case. It's renamed (and so
+  typechecked) the same as ``%One``, even if it appears somewhere that linear
+  modifiers aren't expected. If a user does want the modifier ``1 :: Nat``, they
+  can write it as ``%01``. (Requires ``-XDataKinds``.)
+
+* The linear arrow ``a ⊸ b`` has the same meaning as ```a %1 -> b``. Other
+  modifiers are accepted: ``a %Matchable ⊸ b`` has the same meaning as
+  ``a %Matchable %1 -> b``.
+
+With ``-XNoLinearTypes``, the ``%1`` modifier is not special. It refers to the
+type ``1 :: Nat`` and requires ``-XDataKinds``.
+
+``-XLinearTypes`` implies ``-XModifiers``. But the latter can be explicitly
+disabled with ``-XLinearTypes -XNoModifiers``. This introduces backwards
+compatible behavior:
+
+* Only ``Multiplicity`` modifiers are permitted, and only in the places they're
+  recognized. Any use of a modifier is an error.
+
+* The kind of a modifier is determined by checking for ``Multiplicity``, not
+  through synthesis. So ``Int %m -> Bool`` is forbidden with ``-XLinearTypes
+  -XModifiers``, because ``m`` has unknown kind. But it's permitted with
+  ``-XLinearTypes -XNoModifiers``, equivalently to ``Int %(m :: Multiplicity) ->
+  Bool``.
+
+This behavior may be deprecated in future.
+
+Renaming and typechecking
+-------------------------
```

Comment:
I think this is what @goldfirere [suggested](https://gitlab.haskell.org/ghc/ghc/-/issues/22624#note_469566) to handle this problem.

Reply:
Certain modifiers (that is, modifiers of certain kinds) apply at renamer-time. We can exhaustively list them. (I don't have the full list in my head at the moment, but GHC will always have a fixed set of kinds that can describe modifiers, and some subset of that list affect the renamer.) So that's the specification. If it turns out that the type-checker encounters a modifier whose kind works out to be one of the RN-modifiers (but the modifier wasn't spotted by the renamer, because it required type inference and/or type family/synonym reduction), error.

An alternative approach is to have TC-modifiers use `%`s and RN-modifiers use `&`s, say. Maybe that's a better user experience? I tend to think not, especially because a type-checker-dependent attempt to use an RN-modifier will produce a very clear error message.